src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -51,11 +51,44 @@ internal static class Utf8Encoder
         /// <param name="bytesConsumed">On exit, contains the number of code points that were consumed from the UTF-16 character span.</param>
         /// <param name="charactersWritten">An output parameter to store the number of characters written to <paramref name="utf16"/></param>
         /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
-        public static bool TryDecode(ReadOnlySpan<byte> utf8, Span<char> utf16, out int bytesConsumed, out int charactersWritten)
+        public static unsafe bool TryDecode(ReadOnlySpan<byte> utf8, Span<char> utf16, out int bytesConsumed, out int charactersWritten)
         {
             bytesConsumed = 0;
             charactersWritten = 0;
 
+            if (Vector.IsHardwareAccelerated)
+            {
+                fixed (byte* outerUtf8Ptr = &utf8.DangerousGetPinnableReference())
+                fixed (char* outerUtf16Ptr = &utf16.DangerousGetPinnableReference())
+                {
+                    byte* utf8Ptr = outerUtf8Ptr;
+                    char* utf16Ptr = outerUtf16Ptr;
+
+                    int maxLoops = Math.Min(utf8.Length, utf16.Length) / Vector<byte>.Count;
+                    while (maxLoops != 0)
+                    {
+                        var vec = Unsafe.Read<Vector<byte>>(utf8Ptr);
```

Comment:
FYI, this goes along with my suggestion above to not use fixed and pointer and instead use by-ref.